### PR TITLE
Adding data_info field to models

### LINKED ISSUE #89: Include video size and normalisation information in model metadata
> Works for all pretrained models now. Example command:
> 
> ```
> ./scripts/visualize_model_neurons.py --model_path models/hoefling_2024_GRU_high_res_model.ckpt --save_folder vis/hoefling_2024_GRU_high_res_model
> ```
> 
> Changes:
> 
>     * Make plotting and video generation work for grey-scale videos
> 
>     * Optionally infer stimulus shape from filename [1]
> 
>     * Update visualize_model_neurons script to work with all models
> 
> 
> [1] It might be useful to store the input shape in the CoreReadoutModel, too (we currently pass it to the constructor, but do not store it their). If doing so, we would not have to pass it to the visualize_model_neurons script.

Regarding [1]: I was thinking about this already regarding the video normalisation, and now this is one more strong reason to include input information directly in the model metadata. On top of this, do you think we should include also the values of the video used in normalisation? I noticed they are used in a couple of functions in the script.
And should we make these changes now or incorporate this fix for the time being and deal with these problems later?

_Originally posted by @fededagos in https://github.com/open-retina/open-retina/issues/87#issuecomment-2587350963_

---

The following feature is necessary for several reasons:
1. Easier access to video size metadata for MEI generation
2. Consistent normalisation of models trained on more than one video → video normalisation mean and std need to be made hyperparameters and stored inside the model
3. Easier (and necessary) access to video normalisation values when evaluating the model on an unseen clip (of which the normalisation should follow what was used in training)

### PR DESCRIPTION
**Pull request by fededagos**
Resolves #89 by adding the "data_info" field to models and a function to compute data_info from the same information that is fed into the dataloader-generating function.

### CODE REVIEW

File: openretina/data_io/base.py [hunk 1]
```diff
@@ -84,7 +88,11 @@ def get_n_neurons_per_session(responses_dict: dict[str, ResponsesTrainTestSplit]
...
     train: Float[np.ndarray, "channels train_time height width"],
     test: Float[np.ndarray, "channels test_time height width"],
-) -> tuple[Float[np.ndarray, "channels train_time height width"], Float[np.ndarray, "channels test_time height width"]]:
+) -> tuple[
+    Float[np.ndarray, "channels train_time height width"],
+    Float[np.ndarray, "channels test_time height width"],
+    dict[str, float | None],
+]:
     """
     z-score normalization of train and test movies using the mean and standard deviation of the train movie.
```
Comment (thomasZen) [inline openretina/data_io/base.py:98]: Returns the normalized data for the train and the test data, and a dictionary with the mean and std of the train movie?
^ Add sth like this.

File: openretina/data_io/base.py [hunk 2]
```diff
@@ -84,7 +88,11 @@ def get_n_neurons_per_session(responses_dict: dict[str, ResponsesTrainTestSplit]
 def normalize_train_test_movies(
     train: Float[np.ndarray, "channels train_time height width"],
     test: Float[np.ndarray, "channels test_time height width"],
-) -> tuple[Float[np.ndarray, "channels train_time height width"], Float[np.ndarray, "channels test_time height width"]]:
+) -> tuple[
+    Float[np.ndarray, "channels train_time height width"],
+    Float[np.ndarray, "channels test_time height width"],
+    dict[str, float | None],
```
Comment (thomasZen) [inline openretina/data_io/base.py:94]:
```suggestion
    dict[str, float],
```

Don't we always return a float here?

File: openretina/modules/readout/multi_readout.py [hunk 3]
```diff
@@ -74,6 +74,10 @@ def save_weight_visualizations(self, folder_path: str) -> None:
             os.makedirs(readout_folder, exist_ok=True)
             self._modules[key].save_weight_visualizations(readout_folder)
 
+    @property
+    def sessions(self) -> list[str]:
+        return self.readout_keys()
+
```
Comment (thomasZen) [inline openretina/modules/readout/multi_readout.py:80]: We are not good at naming this. Would be best to have this consistent (either always sessions, or always readout_keys).

File: openretina/models/core_readout.py [hunk 4]
```diff
@@ -181,6 +183,7 @@ def __init__(
         dropout_rate: float = 0.0,
         maxpool_every_n_layers: Optional[int] = None,
         downsample_input_kernel_size: Optional[tuple[int, int, int]] = None,
+        data_info: dict[str, int] | None = None,
```
Comment (thomasZen) [inline openretina/models/core_readout.py]:
```suggestion
        data_info: dict[str, Any] | None = None,
```

I think the int is wrong in the type annotation

File: openretina/models/core_readout.py [hunk 5]
```diff
@@ -25,6 +25,7 @@ def __init__(
         learning_rate: float,
         loss: nn.Module | None = None,
         correlation_loss: nn.Module | None = None,
+        data_info: dict[str, int] | None = None,
```
Comment (thomasZen) [inline openretina/models/core_readout.py]:
```suggestion
        data_info: dict[str, Any] | None = None,
```

I think the int is wrong in the type annotation

File: openretina/data_io/base.py (hunk 2 above)
Comment (thomasZen) [reply openretina/data_io/base.py:94]: but I see now why you did it. If it's much work to get the type annotations consistent in lateron code you can keep it like that

**Review by thomasZen — APPROVED** [review]
Looks reasonable, and will make things easier :)

Joining the documentation with type enforcements for the data info object might be good. Maybe something like using https://docs.python.org/3/library/typing.html#typing.TypedDict gives us both, and it says at runtime it's also just a dict.

Anyways, I'll approve and you can decide if that's worth to change.
